ray-operator/controllers/ray/raycluster_controller.go

```diff
@@ -532,14 +532,13 @@ func (r *RayClusterReconciler) reconcilePods(instance *rayv1alpha1.RayCluster) e
 		}
 
 		runningPods := corev1.PodList{}
-		for _, aPod := range workerPods.Items {
-			if (aPod.Status.Phase == corev1.PodRunning || aPod.Status.Phase == corev1.PodPending) && aPod.ObjectMeta.DeletionTimestamp == nil {
-				runningPods.Items = append(runningPods.Items, aPod)
+		for _, pod := range workerPods.Items {
+			// TODO (kevin85421): We also need to have a clear story of all the Pod status phases, especially for PodFailed.
```

Comment:
Currently, we delete PodFailed pods if the reason is Eviction.
No particular story there -- someone complained that the evicted pods aren't deleted, so we did that.
Maybe we should just delete all failed pods?

Reply:
> Currently, we delete PodFailed pods if the reason is Eviction.

https://github.com/ray-project/kuberay/blob/c4201350ef13598490031f936b945565dcde9dd5/ray-operator/controllers/ray/raycluster_controller.go#L407

Based on the provided code snippet, it appears that only the head Pod with a status of `PodFailed` and reason `Evicted` will be deleted. The function `updateLocalWorkersToDelete` filters out the Pods with a status of PodFailed, so the PodFailed workers will be completely excluded from the deletion process. Do I miss anything?

> Maybe we should just delete all failed pods?

I consider to do that. Because the change could have a significant impact on stability, either positive or negative, I will create a separate PR with more careful considerations at a later time. Some points that I need to figure out:

* With GCS fault tolerance enabled, KubeRay will label certain Pods as unhealthy if their probes generate unhealthy events. Then, KubeRay will delete the Pods with the unhealthy label. This behavior may be refactored soon, and it is a bit overlapped with removing `PodFailed` Pods.

* Some users prefer to retain certain unhealthy Pods or Pods in a PodFailed state for troubleshooting purposes. (e.g. #1022)